Add new rpc.Client.WaitHealthy().

petermattis opened the pull request:
Various "kv" and "client" tests set "failfast=1" which reduces the
number of retries to 1. When using the default retry options, this gives
the sender 250ms in which to perform the initial connection and
heartbeat which sometimes fails to occur, especially on CircleCI. Now we
track when an rpc.Client was created and wait up to 5 seconds for the
initial heartbeat.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3428)

<!-- Reviewable:end -->

petermattis committed:
Add new rpc.Client.WaitHealthy() mechanism.

Various "kv" and "client" tests set "failfast=1" which reduces the
number of retries to 1. When using the default retry options, this gives
the sender 250ms in which to perform the initial connection and
heartbeat which sometimes fails to occur, especially on CircleCI.

Now we track when an rpc.Client was created and wait up to 5 seconds for
the first health indication. A health indication is either a failure to
open a connection, a failed heartbeat or a successful heartbeat.

Fixes #2577.

petermattis commented on kv/dist_sender_server_test.go:
```diff
@@ -250,11 +250,13 @@ func TestMultiRangeScanReverseScanInconsistent(t *testing.T) {
+	for i := range sr.Rows {
+		if key := string(sr.Rows[i].Key); keys[i] != key {
+			t.Errorf("expected key %q; got %q", keys[i], key)
+		}
```
This seems to have been a bug in the previous test. AFAICT, we should get 2 rows here and below, but previously (without the new `WaitHealthy()` time) we were getting only 1. Am I missing anything?

tschottdorf replied:
see l237 above. I think you're supposed to get only the older key, not the newer.

petermattis replied:
Ah, I see. Then why did this change affect the behavior of this test locally (and not on CircleCI)? Something funky is going on.

tschottdorf replied:
agreed. Try printing the timestamp at which the read actually happens (for example in `replica_command.go:executeCommand`).

petermattis commented on rpc/client.go:
```diff
@@ -191,6 +193,31 @@ func (c *Client) Healthy() <-chan struct{} {
 
+// WaitHealthy blocks until the channel is healthy or the HealthWait time has
+// expired.
+func (c *Client) WaitHealthy() bool {
```
This probably deserves a test in this package. Consider this PR a work in progress.

tschottdorf commented:
while you're here, there's also the "issue" of having the backoff happen at the initial connection. So we get all of those "client is unhealthy" log messages, plus a noticeable backoff on each first request. I think that's the main reason for `cli` and `client` taking that long. Maybe that's a low-hanging fruit. See https://github.com/cockroachdb/cockroach/issues/2577

petermattis commented:
@tschottdorf Ok, I have #2577 fixed (now takes 0.02s). But I've been unable to figure out what is going on with `TestMultiRangeScanReverseScanInconsistent`. Mind taking a look?

tschottdorf commented:
cool. Just skip the test and create a stub issue so I'll remember tomorrow.

tschottdorf commented on client/rpc_sender.go:
```diff
@@ -84,16 +85,13 @@ func newRPCSender(server string, context *base.Context, retryOpts retry.Options,
 func (s *rpcSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error) {
 	var err error
+	if !s.client.WaitHealthy() {
+		err = fmt.Errorf("failed to send RPC request %s: client is unhealthy", method)
```
no need to assign here?

petermattis replied:
No need. This was copy&paste. I'll change.

tschottdorf commented on rpc/client.go:
```diff
@@ -191,6 +195,46 @@ func (c *Client) Healthy() <-chan struct{} {
+	select {
+	case <-c.Healthy():
+		return true
+	case <-c.closer:
```
don't need the closer if you have `default` and the result is the same.

petermattis replied:
True.

bdarnell commented on rpc/rpc.go:
```diff
@@ -37,5 +41,6 @@ func (c *Context) Copy() *Context {
 		Stopper:      c.Stopper,
 		RemoteClocks: newRemoteClockMonitor(c.localClock),
 		DisableCache: c.DisableCache,
+		HealthWait:   c.HealthWait,
```
This method could be changed so we don't have to keep the list of fields up to date:

````
func (c *Context) Copy() *Context {
    copy := *c
    copy.RemoteClocks = newRemoteClockMonitor(c.localClock)
    return &copy
}
```

petermattis replied:
Done.

bdarnell commented on rpc/client.go:
```diff
@@ -191,6 +195,45 @@ func (c *Client) Healthy() <-chan struct{} {
+	return false
+}
+
+// WaitHealthy blocks until the channel is healthy, a health indication has
```
This comment should emphasize the return value: "WaitHealthy returns the health of the Client. On the first connection of a newly-created Client, WaitHealthy will block for up to Context.HealthWait if its health has not yet been determined."

petermattis replied:
Done.

bdarnell commented on rpc/client.go:
```diff
@@ -191,6 +195,45 @@ func (c *Client) Healthy() <-chan struct{} {
+	// client (i.e. (un)successful heartbeat or failure to open the connection).
+	select {
+	case <-c.healthReceived:
+		return false
```
This can race with the client becoming healthy: if we get `false` from `isHealthy` and it becomes healthy immediately afterwards, it will return false here. This should return `c.IsHealthy()`

petermattis replied:
True. Done.

bdarnell commented:
As I said in #3395, I'm wary of extending the timeout only for the first connection. It seems like a test-only hack that is only valid as long as none of our tests need to reconnect. I think what we really need here is to refactor `rpcSender.Send` to not use a `util.Retry` loop. This would let us select on the health channel instead of polling it, and we wouldn't have to count an initial unhealthiness as a failure.

petermattis commented:
I'm not settled on this approach, but I'm also not at all happy with the current behavior of the code (prior to this change). Requiring a connection+heartbeat within 250ms in tests is asking for (and receiving) flakiness. 

What would it look like for `rpcSender.Send` to not use a `util.Retry` loop? Note that with this change, we don't retry on health failures. We block until the client is known healthy or unhealthy. Perhaps the retry already isn't necessary with the rest of this change.

PS Waiting for a client we know is unhealthy to become healthy when sending an RPC is the cause of #2577. Retrying in such cases is a good way to waste resources. The approach in this PR is akin to the [circuit breaker](http://martinfowler.com/bliki/CircuitBreaker.html) pattern.

bdarnell commented:
>  Note that with this change, we don't retry on health failures. We block until the client is known healthy or unhealthy. Perhaps the retry already isn't necessary with the rest of this change.

We don't retry on clients that are known unhealthy when `Send` begins, but we will retry if they become unhealthy during the call. But in most cases, anything that we want to retry here would be caught by the initial `WaitHealthy` check, so we are effectively removing retries at this level and could safely remove the loop itself (and the failfast option). (Does anything rely on the fact that in non-failfast mode we used to retry indefinitely here? I don't think so).

> What would it look like for rpcSender.Send to not use a util.Retry loop?

The alternate design I had in mind is something like this:
```
deadline := time.After(s.retryOpts.MaxBackoff)
for {
  select {
  case <-deadline:
    return util.Errorf("client failed to become healthy before deadline")
  case <-s.client.Healthy():
    err := s.client.Call(...)
    if err == nil {
       return nil
    }
  }
}
```

This will always give the client time to become healthy, but because it spends its time in the `select` instead of the `time.Sleep` in `retry.Next` it remains responsive to changes in the client's status.

tschottdorf commented:
I understand the objections in the thread, but meanwhile our builds are still flaky, in part due to this fix not going in. Can we merge this as an interim fix and discuss separately?

tbg committed:
avoid rounding errors in Timestamp.String()

petermattis committed:
Remove the retry loop from client.rpcSender.Send().

Removed now useless "failfast" parameter to client.DB.Open().

tbg committed:
fix up TestMultiRangeScanReverseScanInconsistent

it had a loop to work around non-changing timestamps,
but did not actually clean up the old versions, causing
test failures. in fact, cleanup is really hard there, so
we make sure to be more proactive about the choice of
timestamp instead.

tschottdorf commented:
@petermattis fixed the test failure, pushed two commits to this branch.

petermattis commented:
And I pushed a changed which removes the retry loop from `client.rpcSender.Send`. @bdarnell PTAL

tschottdorf commented:
probably want to turn that mergen into a rebase though.

tamird commented on kv/dist_sender_server_test.go:
```diff
@@ -208,6 +210,8 @@ func TestMultiRangeScanReverseScanDeleteResolve(t *testing.T) {
 // using the clock local to the distributed sender.
 func TestMultiRangeScanReverseScanInconsistent(t *testing.T) {
 	defer leaktest.AfterTest(t)
+	t.Skipf("TODO(tschottdorf): fix me")
```
what's up here?

petermattis commented:
@tschottdorf: Yeah, I'll get the history all prettied up before merging.

tamird commented on rpc/client.go:
```diff
@@ -215,6 +259,14 @@ func (c *Client) close() {
+	setHealthReceived := func() {
+		if healthReceived != nil {
+			close(healthReceived)
+			healthReceived = nil
```
doesn't this race with `WaitHealthy()`?

petermattis replied:
Race in what way? If someone is blocked on `c.healthReceived` they will be woken by the close. Note that `c.healthReceived` is never set to nil.

tamird replied:
oh, this is a local variable. Misread that.

petermattis commented:
Btw, I took another look at #3395 which was @tamird's fix for the same problem. That PR takes the approach of allowing `rpcSender.Send` to wait up to 5s for the first connection to be made, but it does not wait for the connection to be healthy. This PR is different in that it waits for the first health indication (unsuccessful connection, (un)successful heartbeat). I prefer the approach in this PR, but wanted to point out the difference.

bdarnell commented on rpc/client.go:
```diff
@@ -191,6 +195,46 @@ func (c *Client) Healthy() <-chan struct{} {
+// WaitHealthy returns the health of the Client. On the first connection of a
+// newly-created Client, WaitHealthy will block for up to Context.HealthWait if
+// its health has not yet been determined.
+func (c *Client) WaitHealthy() bool {
```
This method can be simplified a bit by removing the first phase, and removing the `case <-c.Healthy():` clause from the last `select`. (we never need to call `isHealthy()` except after reading from `c.healthReceived`).

petermattis replied:
Good point. I've made those simplifications.

bdarnell commented:
LGTM

petermattis commented:
Merging this PR as it is better than the current state of the world. I'll follow-up if there are additional concerns or fallout from removing the retry loop.
